constantine/ecdsa/ecdsa.nim

```diff
@@ -0,0 +1,356 @@
+  # No need for `bits2octets`, because the private key is already a valid
+  # scalar in the field `Fr[C]` and thus < p-1 (`bits2octets` converts
+  # `r` bytes to a BigInt, reduces modulo prime order `p` and converts to
+  # a byte array).
+  let privKeyBytes = privateKey.toBytes()
+
+  # Initial values
+  # Step b: `V = 0x01 0x01 0x01 ... 0x01`
+  var v = arrayWith[32](byte 0x01)
+  # Step c: `K = 0x00 0x00 0x00 ... 0x00`
+  var k = arrayWith[32](byte 0x00)
+
+  # Create HMAC contexts
+  var hmac {.noinit.}: HMAC[sha256]
+
+  # Step d: `K = HMAC_K(V || 0x00 || int2octets(x) || bits2octets(h1))`
+  hmac.round(k, k, v, [byte 0x00], privKeyBytes, msgHashBytes)
+  # Step e: `V = HMAC_K(V)`
+  hmac.round(k, v, v)
+  # Step f: `K = HMAC_K(V || 0x01 || int2octets(x) || bits2octets(h1))`
+  hmac.round(k, k, v, [byte 0x01], privKeyBytes, msgHashBytes)
+  # Step g: `V = HMAC_K(V)`
+  hmac.round(k, v, v)
+  # Step h: Loop until valid nonce found
+  while true:
+    # Step h.1 (init T to zero) and h.2:
+    # `V = HMAC_K(V)`
+    # `T = T || V`
+    # We do not need to accumulate a `T`, because we use SHA256 as a hash
+    # function (256 bits) and Secp256k1 as a curve (also 256 big int).
+    hmac.round(k, v, v) # v becomes T
+
+    # Step h.3: `k = bits2int(T)`
+    var candidate: Fr[C]
+    # `fromDigest` returns `false` if the array is larger than the field modulus,
+    # important for uniform sampling in valid range `[1, q-1]`!
+    let smaller = candidate.fromDigest(v)
+
+    if not bool(candidate.isZero()) and smaller:
+      return candidate
+
+    # Step h.3 failure state:
+    # `K = HMAC_K(V || 0x00)`
+    # `V = HMAC_K(V)`
+    # Try again if invalid
+    hmac.round(k, k, v, [byte 0x00])
+    hmac.round(k, v, v)
+
+proc generateNonce(kind: NonceSampler, msgHash, privateKey: Fr[C]): Fr[C] =
+  case kind
+  of nsRandom: randomFieldElement[Fr[C]]()
+  of nsRfc6979: nonceRfc6979(msgHash, privateKey)
+
+proc signMessage*(message: string, privateKey: Fr[C],
+                  nonceSampler: NonceSampler = nsRandom): tuple[r, s: Fr[C]] =
+  ## Sign a given `message` using the `privateKey`.
+  ##
+  ## By default we use a purely random nonce (uniform random number),
+  ## but passing `nonceSampler = nsRfc6979` uses RFC 6979 to compute
+  ## a deterministic nonce (and thus deterministic signature) given
+  ## the message and private key as base.
+  # 1. hash the message in big endian order
+  let h = hashMessage(message)
+  var message_hash: Fr[C]
+  message_hash.fromDigest(h)
+
+  # loop until we found a valid (non zero) signature
+  while true:
+    # Generate random nonce
+    var k = generateNonce(nonceSampler, message_hash, privateKey)
+
+    # Calculate r (x-coordinate of kG)
+    # `r = k·G (mod n)`
+    let r_point = k * G
+    # get x coordinate of the point `r` *in affine coordinates*
+    let rx = r_point.getAffine().x # element of Fp
+    ## XXX: smarter way for this?
+    let r = Fr[C].fromBig(rx.toBig())
```

Comment:
We need to convert the coordinate in `Fp` to a scalar in `Fr`. The "best" way I could come up with is this to big / from big conversion. Is there a cleaner way?

Reply:
This is the best generic conversion.

In theory for secp256k1, you could take `mod r` because Fp/Fr can be both use canonical representation and still have a fast reduction (#445) however at the moment only Fp uses canonical repr and fast reduction while Fr uses Montgomery representation and Montgomery reduction.